frontend/src/citizen-frontend/messages/state.tsx

```diff
@@ -2,43 +2,276 @@
+): MessageThread[] =>
+  threads.map((t) =>
+    t.id === id
+      ? {
+          ...t,
+          messages: t.messages.map((m) => ({
+            ...m,
+            readAt: m.readAt || new Date()
+          }))
+        }
+      : t
   )
 
-  const value = useMemo(
-    () => ({
-      unreadMessagesCount,
-      refreshUnreadMessagesCount
-    }),
-    [unreadMessagesCount, refreshUnreadMessagesCount]
-  )
+export const MessageContextProvider = React.memo(
+  function MessageContextProvider({ children }: { children: React.ReactNode }) {
+    const [account, setAccount] = useState<Result<MessageAccount>>(Loading.of())
+    const loadAccount = useRestApi(getMessageAccount, setAccount)
 
-  return (
-    <HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
-  )
-})
+    const [threads, setThreads] = useState<ThreadsState>(initialThreadState)
+
+    const setMessagesResult = useCallback(
+      (result: Result<Paged<MessageThread>>) =>
+        setThreads((state) =>
+          result.mapAll({
+            loading() {
+              return state
+            },
+            failure() {
+              return {
+                ...state,
+                loadingResult: result.map(() => undefined)
+              }
+            },
+            success({ data, pages }) {
+              return {
+                ...state,
+                threads: [...state.threads, ...data],
+                loadingResult: Success.of(undefined),
+                pages
+              }
+            }
+          })
+        ),
+      []
+    )
+
+    const loadMessages = useRestApi(getReceivedMessages, setMessagesResult)
+    useEffect(() => {
+      if (threads.currentPage > 0) {
+        setThreads((state) => ({ ...state, loadingResult: Loading.of() }))
+        loadMessages(threads.currentPage)
+      }
+    }, [loadMessages, threads.currentPage])
+
+    const loadMoreThreads = useCallback(() => {
+      if (threads.currentPage < threads.pages) {
+        setThreads((state) => ({
+          ...state,
+          currentPage: state.currentPage + 1
+        }))
+      }
+    }, [threads.currentPage, threads.pages])
+
+    useEffect(() => {
+      if (account.isSuccess) {
+        setThreads((state) => ({ ...state, currentPage: 1 }))
+      }
+    }, [account])
+
+    const [replyState, setReplyState] = useState<Result<void>>()
+    const setReplyResponse = useCallback((res: Result<ReplyResponse>) => {
+      setReplyState(res.map(() => undefined))
+      if (res.isSuccess) {
+        const {
+          value: { message, threadId }
+        } = res
+        setThreads((state) => {
+          const [[old], rest] = _.partition(
+            state.threads,
+            (t) => t.id === threadId
+          )
+          return {
+            ...state,
+            threads: [
+              {
+                ...old,
+                messages: [...old.messages, message]
+              },
+              ...rest
+            ]
+          }
+        })
```

Comment:
This could be given a name. After reading 5 times and checking the types, I now _think_ I understand what it does 😅

Reply:
Better? I ditched lodash too, because the performance impact is negligible.